Project1/MoveHelper.cpp

```diff
@@ -142,3 +142,52 @@ std::vector<ID> MoveHelper::getOtherAlliedCellIDs(const Unit& unit) const{
 }
 	////////////////////////////////////////////////////////////////////////////////
 
+MovementPath MoveHelper::getShortestPath(const Unit& unit, ID destination) {
+	return MovementPath(unit, _map.getShortestPathsMap(unit), _map[destination]);
+}
+
+//TODO(Torrey): optomize this
+MovementPath& MoveHelper::reRoutePath(MovementPath& path, ID destination) {
```

Comment:
When does this get called? It isn't clear to me what it's doing. Is destination the new destination that it's getting rerouted to?

Reply:
This is complicated, I'll explain it tonight.

Reply:
Basically, when you click on a unit, it shows every space it can reach, and will make a path with only a head.  You then use this algorithm to 'make' the paths. What it essentially does is try to use as much of the current path as possible when you try to append cells to the path.

1) When you hover over a tile, if it unreachable, the path remains unchanged.

2) If it is adjacent, reachable, and directly adding it to our current path does NOT exceed our max move, we push_back() that tile.

3) If it is not adjacent to the end of our current path, but can still reach it using our current path as a base, it appends the shortest path to that tile from the back of our current path to the end of our path.

4) If that tile is already in our path, we trim back to it, NO LOOPS.

5) And finally, If the tile is reachable, but there is no way for our current path to reach it without exceeding our max move, we pop off the back cell and attempt rule 4 (above). If that doesn't work, we pop again, and again,and again... until it eventually HAS TO, because we already determined there was SOME path to get there.